akka-typed/src/main/scala/akka/typed/internal/Restarter.scala

```diff
@@ -0,0 +1,272 @@
+    if (restartCount >= 30) // Duration overflow protection (> 100 years)
+      maxBackoff
+    else
+      maxBackoff.min(minBackoff * math.pow(2, restartCount)) * rnd match {
+        case f: FiniteDuration ⇒ f
+        case _                 ⇒ maxBackoff
+      }
+  }
+
+  case object ScheduledRestart
+  final case class ResetRestartCount(current: Int) extends DeadLetterSuppression
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class BackoffRestarter[T, Thr <: Throwable: ClassTag](
+  initialBehavior: Behavior[Any], override val behavior: Behavior[Any],
+  strategy: Backoff, restartCount: Int, blackhole: Boolean) extends Supervisor[Any, Thr] {
+
+  // TODO using Any here because the scheduled messages can't be of type T.
+  //       Something to consider is that timer messages should typically not be part of the
+  //       ordinary public message protocol and therefore those should perhaps be signals.
+  //       https://github.com/akka/akka/issues/16742
+
+  import BackoffRestarter._
+
+  override def loggingEnabled: Boolean = strategy.loggingEnabled
+
+  override def receiveSignal(ctx: ActorContext[Any], signal: Signal): Behavior[Any] = {
+    if (blackhole) {
+      ctx.system.eventStream.publish(Dropped(signal, ctx.self))
+      Behavior.same
+    } else
+      super.receiveSignal(ctx, signal)
+  }
+
+  override def receiveMessage(ctx: ActorContext[Any], msg: Any): Behavior[Any] = {
+    // intercept the scheduled messages and drop incoming messages if we are in backoff mode
+    msg match {
+      case ScheduledRestart ⇒
+        // actual restart after scheduled backoff delay
+        val restartedBehavior = Restarter.initialUndefer(ctx, initialBehavior)
+        ctx.schedule(strategy.resetBackoffAfter, ctx.self, ResetRestartCount(restartCount))
+        new BackoffRestarter[T, Thr](initialBehavior, restartedBehavior, strategy, restartCount, blackhole = false)
+      case ResetRestartCount(current) ⇒
+        if (current == restartCount)
+          new BackoffRestarter[T, Thr](initialBehavior, behavior, strategy, restartCount = 0, blackhole)
+        else
+          Behavior.same
+      case _ ⇒
+        if (blackhole) {
+          ctx.system.eventStream.publish(Dropped(msg, ctx.self))
+          Behavior.same
+        } else
+          super.receiveMessage(ctx, msg)
+    }
+  }
+
+  override def handleException(ctx: ActorContext[Any], startedBehavior: Behavior[Any]): Catcher[Supervisor[Any, Thr]] = {
+    case NonFatal(ex: Thr) ⇒
+      log(ctx, ex)
+      // actual restart happens after the scheduled backoff delay
+      try Behavior.interpretSignal(behavior, ctx, PreRestart) catch { case NonFatal(_) ⇒ }
+      val restartDelay = calculateDelay(restartCount, strategy.minBackoff, strategy.maxBackoff, strategy.randomFactor)
+      ctx.schedule(restartDelay, ctx.self, ScheduledRestart)
+      new BackoffRestarter[T, Thr](initialBehavior, startedBehavior, strategy, restartCount + 1, blackhole = true)
```

Comment:
backoff can probably have many more features, such as buffering, vacation message, and such, but that can be fun for external contributors to do. They did a lot around the classic backoff.